### DISCUSSION TITLE
Relics from C++ in C#: The pointer notation

### DISCUSSION THREAD

**Discussion by ghost** [dotnet/csharplang#6908] [Language Ideas]
Dear all,
one more relic from C++ is the pointer `*` symbol. The first step fixing C++ mistakes is  introduced in C# by adopting the syntax:
```C#
double* p1, p2, p3;
```
as opposed to the C++
```C#
double *p1, *p1, *p2
```
which is very confusing. But why should we use in a new language like C# the `*` at all to begin with. OK C# has the `C` from C++ but why should it inherit its syntactic mistakes?

Wouldn't it be better if instead a more readable syntax was adopted, something like:

```C# 
pointer double p1, p2, p3;
or
double pointer p1, p2, p3;
```

Another school of thought says that pointers are nothing else than hexadecimal memory addresses. As such one would even consider something like:
```C# 
pointer p;
double a  = 0.23;
int n = 100000;
p = address of a;
or
p => a;
or 
p.address = address(a);
or 
p points to a
p.value = 0.25;
p.address = addres(n)
p.value = 1;
```
Would that result in a much more readable code, free of relics such as `*` ? Why should the multiplication symbol be used for indicating pointers or memory addresses? At the end of the day pointers are really memory addresses and they are indeed another type of its own.

**Comment by IS4Code** [comment]
Why not, when we also have arrays denoted using `[]` and nullables using `?`? It is consistent with other syntax constructs that operate on types (as opposed to `ref` which, at the moment still, operates on declarations).

**Reply by ghost** [reply]
Because `[]` is reminiscent of matrices or vectors even on mathematics and the symbol itself implies vectors since it is widely adopted in Linear Algebra. For `?` I have nothing to say it is a question of whether or not something is null and it is easy to remember that it "asks" if something is null since it is a question mark. But, `*` it is widely adopted to indicate multiplication. Why should we use an arithmetic symbol such as `+,-,/,*` for a pointer type?

**Reply by 333fred** [reply]
I mean, I wouldn't necessarily disagree with you for a new language (though it can certainly be argued that * looks like a point, which you follow to pointer). But C# is a 20 year old language with possibly billions of lines of existing code written. I'm not really seeing the motivation to change fundamental C# 1.0 syntax here, especially when pointers are so infrequently used (and new language features are added to further reduce their usage).

**Reply by CyrusNajmabadi** [reply]
> But, * it is widely adopted to indicate multiplication.

And in programming it is used for dereferencing.  We do not slavishly follow mathematics. It's unclear what this solves, and it seems to fall into the same bucket of taking something that doesn't have a problem (or group of users complaining) and just coming up with redundant equivalent syntax for it. 

We've had this syntax for 20 years now, with people fine with it. Changing it now, especially with billions of lines of code out there all consistently using these forms seems like a huge waste of time.

**Reply by ghost** [reply]
Guys I totally agree with both of you. This is why I never suggested to remove this notation for backwards compatibility. Just to "promote" a more readable notation in case you think that would improve C#'s readability and maintainability.  Thank you for your feedback.

**Reply by HaloFour** [reply]
Adding a second new syntax that offers zero other benefits only increases the complexity of the language and the cognitive overhead of every developers using the language that now has to recognize that there are two syntaxes that mean the exact same thing.

**Reply by ghost** [reply]
> Adding a second new syntax that offers zero other benefits only increases the complexity of the language and the cognitive overhead of every developers using the language that now has to recognize that there are two syntaxes that mean the exact same thing.

This is called progress. Other languages allow more advanced syntax if it promotes readability keeping backward compatibility.

**Reply by DavidArno** [reply]
I've been using C# on and off for 20 years or so. During that time, I've had to use `unsafe` probably less than a dozen times. I'm not sure if I ever had to use the pointer notation on those occasions. This is fairly typical for most C# developers: pointers are an incredibly niche feature. Having the language team implement a second way of denoting pointers and then having reams of documentation around the language updated to record this isn't progress: it would be a monumental waste of everyone's time. Time that could be better spent on addressing missing features from the language or improvements to features that are used all the time by most developers.

**Reply by theunrepentantgeek** [reply]
> "promote" a more readable notation

We have over 50 years, as an industry, of familiarity with C style pointer syntax. (C is 51 years old, and I believe the syntax is older still.)

Virtually every other language with pointer support uses this syntax or something very close to it, allowing developers to transfer that knowledge to new domains. 

Introducing unique syntax into C# to do things in a wholly different way doesn't promote readability or clarity - it would do the exact opposite.

**Reply by ghost** [reply]
then why didn't you keep the exact same C/C++ pointer syntax?
```C#
double *a1, *a2 *a3;
```
Because you realized it is wrong. You should have thought one more step forward back then. It does not hurt to do it now if you agree since you have introduced so many other keywords for several different things, and it will be straight forward, no much to document introducing `pointer` as alternative to `*`. One more keyword readable, self explanatory. I understand that functionality such as C++ templates may take years, but this? Come on. It is way easier. And people will love it.

**Reply by DavidArno** [reply]
> then why didn't you keep the exact same C/C++ pointer syntax? ... Because you realized it is wrong ...

That is very likely a false statement. The semantics are completely different so the syntax needs to be different. In your C code, you are specifying that `a1` is a pointer to the `double` type. But you can't do that in C#. Pointers are types themselves. So it has to be expressed as `double*`: it is declaring a "pointer to double" type and then `a1` etc are declared as being of that type. 

> And people will love it.

People will be oblivious to it as almost no one uses pointer types. If you are using them, you are more than likely doing things wrongly.

**Reply by ghost** [reply]
You are not using it, because most likely you are not calling external libraries such as MKL or other performance libraries written in Fortran. However, there are thousands of people who need to interface with libraries written in other Languages such as Fortran or C++ for performance reasons.

**Reply by HaloFour** [reply]
Ultimately the team is not going to add another syntax one that has existed in this language for over 20 years and in other languages for over 50 years simply because you prefer a keyword.  The resemblance to C++ is intentional and desired, anyone writing performance-oriented code is already very familiar with it.  The time and energy that would be involved to implement this suggestion would be better spent literally anywhere else.

**Comment by tannergooding** [comment]
Also worth noting that `T**` and `T***` and further are all possible

`pointer T` is more verbose and arguably less readable, as is `pointer pointer T` and other deeper levels of indirection.

**Reply by ghost** [reply]
Sorry, to have to say this, but for a new guy to programming:

`pointer T` is clear that the type is pointer. It is not clear what this is `T*` ; multiplication ? Just try to recall you impression the first time you read the pointers chapter of a C/C++ book. Didn't you blinked twice "why in the world they are using multiplication for indicating a type?"

**Reply by HaloFour** [reply]
For someone new to programming they'd probably wonder more why an asterisk is used for multiplication instead of the accepted mathematical operators `×`, `⋅` or by juxtaposition.  The use of asterisk here is very specific to programming languages.

But, like all things, anyone new to a language is going to have to learn how that language works.  The word "pointer" doesn't teach them anymore more about pointers than a symbol does.

**Reply by ghost** [reply]
Still it is awkward to use `*` that you use in `5*30` in order to define a pointer data type. Usually data types are defined with whole names.

**Reply by HaloFour** [reply]
No more awkward than anything else a new programmer would have to learn.  Every language is full of its own idioms and idiosyncrasies.  But of the popular languages that do expose pointers it does seem that the C-like notation is, by far, the most common, so using a different form of notation would be more confusing.  There's a reason that so many languages use that notation.

**Reply by FaustVX** [reply]
More importantly, in C#, if you are a beginner, just don't use pointer, you don't have to.
And the time you need pointer in your code, you are probably aware of that syntax, so you'll recognize it easily.

**Reply by ghost** [reply]
OK, guys do not implement it, if it is such a big project for you. Keep the `*` since it promotes clarity. At the end of the day C# is another C derivative.

**Reply by CyrusNajmabadi** [reply]
> Usually data types are defined with whole names.

This is definitely not true. We don't do that for arrays, nullable ref types, named tuples, etc.  We have shorthand syntax because we believe strongly that types are common enough and special enough to warrant it. 

Furthermore, as mentioned in other posts, we have never used keyword names to augment a type.  So 'pointer' (which is the least used type of all of them) would now be distinctly different from everything else. That's not a good thing.

**Reply by ghost** [reply]
You are forgetting that `pointer` is a type of its own. It is a hexadecimal address. Nothing else and nothing more. It is wrong by design of several languages including C,C++,C#,ADA,Fortran, and many others to associate it with a type just to allow pointer arithmetics and other crap that result in memory bugs such as `segmentation faults`. It is totally wrong. Pointers show somewhere in memory. If you want to read something from memory pointed by a pointer you need of course to know the type of data pointed by the pointer. But this can be recognized by the compiler at compile time, by the assignment of the pointer to some address of data. All languages have got that wrong. This is where your confusion emanates from. Forget what other people have done in the past and think how would you do that anew. Do not repeat mistakes of the past. Otherwise you are stack in the same pitfall as your ancestors. Learn by their mistakes; do not repeat them.

**Reply by CyrusNajmabadi** [reply]
> You are forgetting that pointer is a type of its own.

We're not forgetting anything.  You're literally talking to the language designers here.  We are *very* well aware of what these constructs are, as:

1. a language feature.
2. a runtime feature.
3. a low-level computer hardware representation.

Please stop treating us as children who are somehow blind to some aspect of this space that you are informing us of.  

> All languages have got that wrong.

Nothing about your suggestion changes teh design of pointers at all, so this tangent of yours isn't helping your case at all.  Your suggestion here is literally about changing the syntax here, nothing more.  Pointers would still work exactly as they always have, just with a more verbose way to have to write them out.

> This is where your confusion emanates from. 

@Dhmosthenis I'm going to state this as clearly as possible.  If you persist in this style of argument, you are going to find no willingness from the language designers or community to engage with your ideas.  We are all experienced professionals here, with literal decades under our belts working in the space of language design.  We're happy to have healthy and constructive discussions about potential future changes in the language.  We're not going to engage with people who are not willing to be constructive, and view their view as the only right way forward, and anything that might be contrary to that to simply be a mistake and nothing more.

**Comment by bernd5** [comment]
Btw.: Bjarne Stroustrup (the founder of C++) very much dislikes the C pointer declarator syntax (and he declares usually only a single pointer on one line). He useses: `Type* declarator` instead of `Type *declarator`.
To be compatible with C - C++ inherited that style, but...
